src/week06/tests/test_negative_r_timed.py

```diff
@@ -0,0 +1,90 @@
+import hypothesis
+import hypothesis.strategies as st
+from opshin import build
+
+from src.utils import network
+from src.utils.mock import MockChainContext, MockUser
+from src.week06.lecture.negative_r_timed import validator, CustomDatum
+import opshin.prelude
+import datetime
+import pycardano
+from src.week06 import lecture_dir
+
+
+def make_mock_context(
+    valid_range: opshin.prelude.POSIXTimeRange,
+):
+    tx_info = opshin.prelude.TxInfo(
+        inputs=None,
+        reference_inputs=None,
+        outputs=None,
+        fee=None,
+        mint=None,
+        dcert=None,
+        wdrl=None,
+        valid_range=valid_range,
+        signatories=None,
+        redeemers=None,
+        data=None,
+        id=None,
+    )
+    context = opshin.prelude.ScriptContext(tx_info=tx_info, purpose=None)
+    return context
+
+
+@hypothesis.given(dt=st.datetimes(), redeemer=st.integers())
+def test_property_before_fails(dt: datetime.datetime, redeemer: int):
+    deadline: opshin.prelude.POSIXTime = int(
+        dt.timestamp() * 1000
+    )  # plutus contracts uses milliseconds
+    run(CustomDatum(deadline), redeemer)
+
+
+def setup_user(context: MockChainContext):
+    user = MockUser(context)
+    user.fund(10000000)  # 10 ADA
+    return user
+
+
+def run(datum, redeemer_data, *args):
+    mock_chain_context = MockChainContext()
+    # SETUP USERS
+    u1 = setup_user(mock_chain_context)
+    u2 = setup_user(mock_chain_context)
+    # build script
+    plutus_script = build(lecture_dir.joinpath("negative_r_timed.py"), *args)
+    script_hash = pycardano.plutus_script_hash(plutus_script)
+    script_address = pycardano.Address(script_hash, network=network)
+
+    # USER 1 LOCKS 100 Lovelaces ("val") IN VALIDATOR
+    val = pycardano.Value(coin=2000000)  # 2 ADA
+    tx_builder = pycardano.TransactionBuilder(mock_chain_context)
+    tx_builder.add_input_address(u1.address)
+    tx_builder.add_output(
+        pycardano.TransactionOutput(
+            script_address, amount=val, datum=datum, script=plutus_script
+        )
+    )
+    tx = tx_builder.build_and_sign([u1.signing_key], change_address=u1.address)
+    mock_chain_context.submit_tx_mock(tx)
+
+    # WAIT FOR A BIT
+    mock_chain_context.wait(1000)
+
+    # USER 2 TAKES "val" FROM VALIDATOR
+    utxo = mock_chain_context.utxos(str(script_address))[0]
+    tx_builder = pycardano.TransactionBuilder(mock_chain_context)
+    tx_builder.add_script_input(
+        utxo,
+        redeemer=pycardano.Redeemer(
+            redeemer_data, pycardano.ExecutionUnits(1000000, 1000000)
```

Comment:
This number of steps is much too little! Contrast this to the default max tx memsteps of `14000000` and max tx cpu steps of `10000000000`.

What happens usually in the tx builder is that it first submits the transaction to `evaluate_tx`, which computes the required steps (starting off with the allowed maximum) and then fixes the execution units.

Reply:
I wanted to do it with `evaluate_tx`, but at first it seemed complicated to do with `pyaiken`. Could we get all information we need for `uplc.eval` with just a `Transaction` object?

Reply:
I am working on something :)

Reply:
Nice, let me know :)

Reply:
Pushed some stuff and working on ironing out bugs

Reply:
Sounds good, I'll check back in a few